_source/_posts/2021-02-26-unity-webgl-playfab-authorization.md

```diff
@@ -0,0 +1,654 @@
+
+Finally, `Authenticate()` referenced a few helper functions for generating the code challenge\verifier, encryption, and encoding.
+
+```csharp
+/// <summary>
+/// Returns URI-safe data with a given input length.
+/// </summary>
+/// <param name="length">Input length (nb. output will be longer)</param>
+/// <returns></returns>
+public static string randomDataBase64url(uint length)
+{
+        RNGCryptoServiceProvider rng = new RNGCryptoServiceProvider();
+        byte[] bytes = new byte[length];
+        rng.GetBytes(bytes);
+        return base64urlencodeNoPadding(bytes);
+}
+
+/// <summary>
+/// Returns the SHA256 hash of the input string.
+/// </summary>
+/// <param name="inputString"></param>
+/// <returns></returns>
+public static byte[] sha256(string inputString)
+{
+        byte[] bytes = Encoding.ASCII.GetBytes(inputString);
+        SHA256Managed sha256 = new SHA256Managed();
+        return sha256.ComputeHash(bytes);
+}
+
+/// <summary>
+/// Base64url no-padding encodes the given input buffer.
+/// </summary>
+/// <param name="buffer"></param>
+/// <returns></returns>
+public static string base64urlencodeNoPadding(byte[] buffer)
+{
+        string base64 = Convert.ToBase64String(buffer);
+
+        // Converts base64 to base64url.
+        base64 = base64.Replace("+", "-");
+        base64 = base64.Replace("/", "_");
+        // Strips padding.
+        base64 = base64.Replace("=", "");
+
+        return base64;
+}
+```
+
+## Log Into PlayFab
+
+### Configure PlayFab OIDC
+
+Before PlayFab can accept authorization tokens from Okta, a new [Open ID Connection](https://docs.microsoft.com/en-us/rest/api/playfab/admin/authentication/createopenidconnection?view=playfab-rest) will need to be created. This is the connection that will give PlayFab the needed information to validate the player's ID Token.
+
+Log into your PlayFab portal and click on the title for your project. If you do not have a `Studio` or `Title` created, create them now. The studio and title would have been needed for the PlayFab Unity SDK setup.
+
+{% img blog/unity-webgl-playfab-authorization/12_PlayFabTitle.png alt:"PlayFab Title" width:"800" %}{: .center-image }
+
+Next, click the gear next to the title name in the top left and select **Title Settings**.
+
+{% img blog/unity-webgl-playfab-authorization/13_PlayFabGear.png alt:"PlayFab Gear" width:"800" %}{: .center-image }
+
+Click the `Open ID Connect` tab and select **New Connection**
+
+{% img blog/unity-webgl-playfab-authorization/14_PlayFaboidc.png alt:"PlayFab OIDC" width:"800" %}{: .center-image }
+
+In the `New Connection` menu, name the Connection ID `Okta` and enter the `Client ID` and `Issuer` from the app set up by the Okta CLI. `Client Secret` is not used because of `PKCE `, so enter any random string into the `Client Secret` box in PlayFab. Finally, check `Ignore nonce` and click **Save Connection**.
```

Comment:
Perhaps explain why the nonce is being ignored?

Reply:
Because its not actually OIDC its just validating an ID token. :D I added something.